Sparkc 94 - Remove Thrift dependency - replace use of describe_splits_ex with CQL access

pkolaczk opened the pull request.

jacek-lewandowski commented:
@pkolaczk I haven't finished the review yet, but one thing I'd suggest is to revert the name change from `nativePort` to `port`. This value is still the native port and it is named `native` in `cassandra.yaml` file. Other than that, I thing this is unnecessary change in the set of our configuration properties. Would you consider leaving `nativePort`?

pkolaczk commented:
It is named native in cassandra.yaml, because Cassandra supports both thrift (rpc) and native interface. However, it is named just "port" in the official DataStax drivers (see `ClusterBuilder` API). I left "native" in the docs.

sdelmas commented:
retest this please

sdelmas commented:
Piotr, I hope it's ok to use this one to test our PR integration real quick.

sdelmas commented:
retest this please

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraTableRowReaderProvider.scala:
```diff
@@ -29,13 +29,13 @@ trait CassandraTableRowReaderProvider[R] {
   protected def readConf: ReadConf
 
-  protected def fetchSize: Int = readConf.fetchSize
+  protected def fetchSize: Int = readConf.fetchSizeInRows
```
We are using size in bytes, megabytes, rows in more places (like `BatchSize`). Why don't create something like trait `Size` with subclasses `SizeInBytes` and `SizeInRows`, and then use these types where appropriate?

pkolaczk replied:
That's a good idea but even a better idea would be to encode the unit in the type. And even "more better" idea is to not code it by ourselves, but use something already invented for this purpose e.g. squants http://www.squants.com/. However, we should use it in all places and not just this one, so let's do it in a separate ticket.

jacek-lewandowski replied:
sgtm

jacek-lewandowski commented on spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/Murmur3PartitionerTokenRangeSplitter.scala:
```diff
@@ -2,29 +2,33 @@ package com.datastax.spark.connector.rdd.partitioner
         new LongToken(r),
         range.replicas,
-        Some((estimatedRows / n).toInt))
+        Some((rangeSize / n).toInt))
```
So, is the `rowCount` field name in `TokenRange` still valid?

pkolaczk replied:
Indeed, it should be named differently.

jacek-lewandowski commented:
I have a similar concern as @artem-aliev does - what about older Cassandra versions? If I understand it correctly, it won't split the token ranges at all in that case, right? If so, it may lead to OOMs for large datasets. Maybe we could consider one of the following ideas:
- add an optional parameter to the read configuration, for example `sparkPartitionsPerTokenRange`, which would allow to explicitly specify how to split each token range
- create a pluggable mechanism for Thrift - I mean, we could provide the old functionality as a separate module and therefore not include Thrift dependency to the core

besides, LGTM - please consider my comments if you like (one more thing, which I couldn't comment in the code is that `CassandraRDDPartitioner.TokenRangeSampleSize` seems to be unused now).

pkolaczk commented:
Yes, I'm working on an optional parameter to set the number of partitions explicitly. It will be useful also if someone doesn't like the automatic split-size tuning or for some other reasons it is not accurate.

jacek-lewandowski commented:
retest this please

jacek-lewandowski commented:
retest this please

jacek-lewandowski commented:
retest this please

pkolaczk commented:
retest this please

pkolaczk commented:
retest this please

pkolaczk commented:
retest this please

pkolaczk commented:
```
		::::::::::::::::::::::::::::::::::::::::::::::
		::          UNRESOLVED DEPENDENCIES         ::
		::::::::::::::::::::::::::::::::::::::::::::::
		:: org.scala-sbt#sbt;0.13.8: not found
```

WTF?

pkolaczk commented:
@bcantoni @helena can you look why it has problems downloading sbt?

pkolaczk commented:
retest this please

pkolaczk commented:
Rebased onto master.

pkolaczk commented:
Rebased onto master.

pkolaczk commented:
retest this please

pkolaczk commented:
@jacek-lewandowski can you please review the last two commits? I added setting fixed split count for those who are running on older Cassandra and I added fetching listen addresses from the peers table to improve data locality in cloud environments.

jacek-lewandowski commented:
+1
